### PR TITLE
Clientprefs helpers for integers and strings

### PR DESCRIPTION
**Pull request by eyal282**
Title

### CODE REVIEW

**Comment by MAGNAT2645** [comment]
1) You shouldn't use `MarkNativeAsOptional` because these methods are not natives.
2) `Cookie.Get` returns empty string if there's no cookie stored for specific client. `StringToInt` in this case will return `0` and you can't really tell if that's the real value stored or that's the default value returned by `StringToInt` due to invalid (non-numeric or empty) string. I think it should be useful to provide default value if there's no cookie stored. For example:
```sourcepawn
	public int GetInt(int client, int defaultValue = 0) {
		char sValue[11];
		this.Get(client, sValue, sizeof(sValue));

                int value;
                if (!StringToIntEx(sValue, value))
                       value = defaultValue;
		return value;
	}

	public float GetFloat(int client, float defaultValue = 0.0) {
		char sValue[11];
		this.Get(client, sValue, sizeof(sValue));

                float value;
                if (!StringToFloatEx(sValue, value))
                       value = defaultValue;
		return value;
	}
```

**Comment by eyal282** [comment]
> 1. You shouldn't use `MarkNativeAsOptional` because these methods are not natives.
> 2. `Cookie.Get` returns empty string if there's no cookie stored for specific client. `StringToInt` in this case will return `0` and you can't really tell if that's the real value stored or that's the default value returned by `StringToInt` due to invalid (non-numeric or empty) string. I think it should be useful to provide default value if there's no cookie stored. For example:
> 
> ```sourcepawn
> 	public int GetInt(int client, int defaultValue = 0)
> 	{
> 		char buffer[11];
> 		this.Get(client, buffer, sizeof(buffer));
> 
>                 int value;
>                 if (!StringToIntEx(buffer, value))
>                 {
>                        value = defaultValue;
>                 }
> 		return value;
> 	}
> 
> 	public float GetFloat(int client, float defaultValue = 0.0)
> 	{
> 		char buffer[11];
> 		this.Get(client, buffer, sizeof(buffer));
> 
>                 float value;
>                 if (!StringToFloatEx(buffer, value))
>                 {
>                        value = defaultValue;
>                 }
> 		return value;
> 	}
> ```

I made the suggested changes you gave.

**Comment by MartLegion** [comment]
In case this is approved, would be nice having something like that for DispatchKeyValue as well

File: plugins/include/clientprefs.inc [hunk 1]
```diff
@@ -141,6 +166,42 @@ methodmap Cookie < Handle {
...
+	//
+	// @param client        Client index.
+	// @return              Integer value of cookie
+	// @error               Invalid cookie handle or invalid client index.
+	public int GetInt(int client, int defaultValue = 0)
+	{
+		char buffer[11];
+		this.Get(client, buffer, sizeof(buffer));
+
+                int value;
```
Comment (peace-maker) [inline plugins/include/clientprefs.inc:179]: Mix of tabs and spaces.

File: plugins/include/clientprefs.inc [hunk 2]
```diff
@@ -141,6 +166,42 @@ methodmap Cookie < Handle {
...
+	}
+	
+	// Retrieve the float value of a Client preference cookie.
+	//
+	// @param client        Client index.
+	// @return              Float value of cookie
+	// @error               Invalid cookie handle or invalid client index.
+	public float GetFloat(int client, float defaultValue = 0.0)
+	{
+		char buffer[11];
```
Comment (peace-maker) [inline plugins/include/clientprefs.inc:194]: Is the largest float really only 10 characters long? I wouldn't prematurely optimize buffer sizes and just use some large-enough power of two.

**Review by peace-maker — CHANGES_REQUESTED** [review]
Those seem nice to have. Some comments inline.

File: plugins/include/clientprefs.inc [hunk 3]
```diff
@@ -126,12 +126,37 @@ methodmap Cookie < Handle {
...
+	// @error               Invalid cookie handle or invalid client index.
+	public void SetFloat(int client, float value) {
+		char sValue[11];
+		FloatToString(value, sValue, sizeof(sValue));
+		
+		this.Set(client, sValue);
+	}
+
 
 	// Retrieve the value of a Client preference cookie.
```
[the remark this answers is not in the document]
Comment (MartLegion) [reply plugins/include/clientprefs.inc:161]: SetFloat still small [11]

[the remark this answers is not in the document]
Comment (peace-maker) [reply plugins/include/clientprefs.inc:161]: I'm talking about the comment I've commented on on line 136/161. You changed the docs for `.Set` to say "string value", so the docs for `.Get` could say that as well. It's not that important, sorry for not being clear.

File: plugins/include/clientprefs.inc [hunk 4]
```diff
@@ -132,6 +132,31 @@ methodmap Cookie < Handle {
...
+		this.Set(client, sValue);
+	}
+	
+	// Set the float value of a Client preference cookie.
+	//
+	// @param client        Client index.
+	// @param value         Float value to set.
+	// @error               Invalid cookie handle or invalid client index.
+	public void SetFloat(int client, float value) {
+		char sValue[11];
```
Comment (peace-maker) [inline plugins/include/clientprefs.inc:154]: Same comment about buffer size for large float values applies here too I think.

File: plugins/include/clientprefs.inc [hunk 5]
```diff
@@ -141,7 +166,43 @@ methodmap Cookie < Handle {
...
+	public int GetInt(int client, int defaultValue = 0)
+	{
+		char buffer[11];
+		this.Get(client, buffer, sizeof(buffer));
+
+		int value;
+                if (!StringToIntEx(buffer, value))
+                {
+                       value = defaultValue;
+                }
```
Comment (peace-maker) [inline plugins/include/clientprefs.inc:183]: Still lots of mixed tabs and spaces. I didn't comment on every line, but only wanted to point out the general problem with your changes.

**Review by peace-maker — CHANGES_REQUESTED** [review]

**Review by peace-maker — APPROVED** [review]
Thank you for this!

**Pull request merged**